renderdoc/driver/gl/cgl_platform.mm

```diff
@@ -1,53 +1,423 @@
+
+static Threading::CriticalSection s_ContextIndexesArrayLock;
+static rdcarray<NSOpenGLContext *> s_ContextIndexes;
+static rdcarray<Threading::CriticalSection *> s_ContextLocks;
+#if RD_USE_CONTEXT_LOCK_COUNTS
+static rdcarray<int32_t> s_ContextLocksCount;
+#endif    // #if RD_USE_CONTEXT_LOCK_COUNTS
+
+static void scheduleContextSetView(int contextIndex, NSView *view);
+static void scheduleContextUpdate(int contextIndex);
+
+static void RandomSleep()
+{
+#if RD_THREAD_RANDOM_SLEEP
+  usleep(rand() % 1000);
+#endif    // #if RD_THREAD_RANDOM_SLEEP
+}
+
+static NSOpenGLContext *GetNSOpenGLContext(void *context)
+{
+  NSOpenGLContext *nsglContext = (NSOpenGLContext *)context;
+  RDCASSERT([nsglContext isKindOfClass:[NSOpenGLContext class]]);
+  return nsglContext;
+}
+
+static NSView *GetNSView(void *view)
+{
+  NSView *nsView = (NSView *)view;
+  RDCASSERT([nsView isKindOfClass:[NSView class]]);
+  return nsView;
+}
+
+static void IncrementContextLockCount(int contextIndex)
+{
+#if RD_USE_CONTEXT_LOCK_COUNTS
+  Atomic::Inc32(s_ContextLocksCount.data() + contextIndex);
+#endif    //#if RD_USE_CONTEXT_LOCK_COUNTS
+}
+
+static void DecrementContextLockCount(int contextIndex)
+{
+#if RD_USE_CONTEXT_LOCK_COUNTS
+  Atomic::Dec32(s_ContextLocksCount.data() + contextIndex);
+#endif    //#if RD_USE_CONTEXT_LOCK_COUNTS
+}
+
+static void LockContext(int contextIndex)
+{
+  s_ContextLocks[contextIndex]->Lock();
+  IncrementContextLockCount(contextIndex);
+}
+
+static bool TryLockContext(int contextIndex)
+{
+  if(s_ContextLocks[contextIndex]->Trylock())
+  {
+    IncrementContextLockCount(contextIndex);
+    return true;
+  }
+  return false;
+}
+
+static void UnLockContext(int contextIndex)
+{
+  DecrementContextLockCount(contextIndex);
+  s_ContextLocks[contextIndex]->Unlock();
+}
+
+static void LockWindowIndexesArray()
+{
+  s_WindowIndexesArrayLock.Lock();
+}
+
+static void UnLockWindowIndexesArray()
+{
+  s_WindowIndexesArrayLock.Unlock();
+}
+
+// s_WindowIndexesArrayLock must be locked by the caller
+static bool findWindowIndex(NSView *nsView, int &index)
+{
+  int windowCount = (int)s_WindowIndexes.size();
+  int firstFreeIndex = windowCount;
+  for(int i = 0; i < windowCount; ++i)
+  {
+    if(s_WindowIndexes[i] == nsView)
+    {
+      index = i;
+      return true;
+    }
+    if((s_WindowIndexes[i] == nil) && (i < firstFreeIndex))
+    {
+      firstFreeIndex = i;
+    }
+  }
+  index = firstFreeIndex;
+  return false;
+}
+
+static int getWindowIndex(NSView *nsView)
+{
+  LockWindowIndexesArray();
+  int index = -1;
+  findWindowIndex(nsView, index);
+  int windowCount = (int)s_WindowIndexes.size();
+  if(index < windowCount)
+  {
+    RDCASSERT(index >= 0);
+    RDCASSERT(index < (int)s_WindowIndexes.size());
+    RDCASSERT(index < (int)s_WindowWidths.size());
+    RDCASSERT(index < (int)s_WindowHeights.size());
+    s_WindowIndexes[index] = nsView;
+    s_WindowWidths[index] = 0;
+    s_WindowHeights[index] = 0;
```

Comment:
This will reset the width/height to 0,0 even if the window is found in `findWindowIndex` above. I think if `findWindowIndex` returns true this function should early out and just return the index - i.e. if the view is already in the array. This code should only run in the cases where the view isn't in the array and we're either re-using a slot or adding a new one.

Reply:
Good spot : not sure how it was working with this code, changed to this code which matches how it used to work before findWindowIndex was implemented

static int getWindowIndex(NSView *nsView)
{
  SCOPED_LOCK(s_WindowPtrsArrayLock);
  int index = -1;
  if(findWindowIndex(nsView, index))
  {
    return index;
  }
  const int windowCount = s_WindowPtrs.count();

90e179aebe2381080719784dbccaeee4afa3cbed